### PR TITLE
Format vmmc output

### PR DESCRIPTION
**Pull request by jeffeaton**
This PR formats the DMMPT2 output and appends it to the PEPFAR Datapack CSV if 

It builds on #418. I have made a separate PR in case you want to merge that one immediately before reviewing this, but they can be combined and merged together.

This code does _not_ check that the DMMPT2 file `area_id` are consistent with the Naomi fit.  What do we want to test there, and how do we want it to fail (error vs. warning)?

Options for what to check:

1. All DMPPT2 `area_id` are contained in the Naomi fit `area_id`
2. All Naomi fit `area_id` are contained in the DMPPT2 output (we don't want this option)
3. All Naomi fit `area_id` at the PSNU level are contained in the DMMPT2 output file

I think we want option 3.

I'm not sure how it should fail:
* A hard error could be a nuisance as it stops them from downloading Naomi zip altogether. 
* But a warning likely to be glossed over and produce a Datapack CSV that is missing some rows for the circumcision outputs

One option would be to write the Naomi zip file, but omit the PEPFAR CSV if there is an issue with the DMMPT2 file. (Still could be slightly annoying, but less so than error writing the Naomi zip altogether).

### CODE REVIEW

**Review by r-ash — CHANGES_REQUESTED** [review]
> This code does not check that the DMMPT2 file area_id are consistent with the Naomi fit. What do we want to test there, and how do we want it to fail (error vs. warning)?

Related to this in hintr I am validating that the regions in the uploaded VMMC file. If any region appears in the VMMC file which is not in the shape file, then the VMMC file will raise a validation error. 

> I'm not sure how it should fail:

Writing the zip file but omitting the PEPFAR CSV sounds good to me. We could raise a warning too so that it is clear to users why it is missing. 

You could still create the datapack CSV but without the DMMPT2 indicators, thought that might be more confusing about why they are missing.

If we split the datapack and output zip into a separate downloads down the line it could avoid this pain. We could have a proper error if datapack fails to generate but the output zip would generate fine.

**Comment by jeffeaton** [comment]
> > This code does not check that the DMMPT2 file area_id are consistent with the Naomi fit. What do we want to test there, and how do we want it to fail (error vs. warning)?
> 
> Related to this in hintr I am validating that the regions in the uploaded VMMC file. If any region appears in the VMMC file which is not in the shape file, then the VMMC file will raise a validation error.

Great — I think this is reasonable. And then user has the option to "remove" the VMMC file and proceed with fitting (and VMMC file will be passed as `NULL`)?

> 
> > I'm not sure how it should fail:
> 
> Writing the zip file but omitting the PEPFAR CSV sounds good to me. We could raise a warning too so that it is clear to users why it is missing.
> 
> You could still create the datapack CSV but without the DMMPT2 indicators, thought that might be more confusing about why they are missing.

Let's go with the "create data pack CSV without DMPPT2 indicators + warning" option. This shouldn't really happen if the VMMC `area_id`s are being validated on upload.

> If we split the datapack and output zip into a separate downloads down the line it could avoid this pain. We could have a proper error if datapack fails to generate but the output zip would generate fine.

Yes this is the ideal eventual resolution.

File: tests/testthat/test-pepfar-datapack.R [hunk 1]
```diff
@@ -69,3 +69,18 @@ test_that("datapack export writes correct psnu_level", {
...
+  vmmc_datapack_raw <- openxlsx::read.xlsx(vmmc_path, sheet = "Datapack inputs", startRow = 2)
+  vmmc_datapack <- transform_dmppt2(vmmc_datapack_raw)
+
+  tmpf <- tempfile(fileext = ".csv")
+  res <- write_datapack_csv(a_output_full, tmpf, dmppt2_output = vmmc_datapack)
+
+  datapack1 <- readr_read_csv(res)
+  expect_true(!any(is.na(datapack1)))
+  expect_true(all(c("VMMC_CIRC_SUBNAT.T_1", "VMMC_TOTALCIRC_SUBNAT.T_1") %in%
+                    datapack1$indicator_code))
```
[the remark this answers is not in the document]
Comment (jeffeaton) [reply tests/testthat/test-pepfar-datapack.R:85]: Thanks, done (b82c965)

**Comment by jeffeaton** [comment]
> > > I'm not sure how it should fail:
> > 
> > 
> > Writing the zip file but omitting the PEPFAR CSV sounds good to me. We could raise a warning too so that it is clear to users why it is missing.
> > You could still create the datapack CSV but without the DMMPT2 indicators, thought that might be more confusing about why they are missing.
> 
> Let's go with the "create data pack CSV without DMPPT2 indicators + warning" option. This shouldn't really happen if the VMMC `area_id`s are being validated on upload.

Actually, on further thought — there are a couple of countries (Kenya, Ethiopia) where the DMPPT2 results are only for some districts, not the whole country.  This check might create an issue for those (not sure how the DMPPT2 output file handles these). 

For now, let's just go with your upload check (DMMPT2 `area_id` are all contained in Naomi file) and leave it at that.

**Comment by jeffeaton** [comment]
Thanks 2efe848 should fix the broken test. This is ready for re-review.

File: R/pepfar-datapack.R [hunk 2]
```diff
@@ -49,15 +52,19 @@ write_datapack_csv <- function(naomi_output,
...
-    ## If using the demo data, don't print this warning.
-    ## The demo data only contain levels 0:2, but have ISO3 = "MWI", for which psnu_level = 3.
-    ## A better way to handle this is to change the ISO3 for the demo data to an artificial
-    ## code. However, some hint validation needs to be relaxed to enable this.
-    if (!("MWI_2_5_demo" %in% naomi_output$meta_area$area_id && setequal(0:2, naomi_output$meta_area$area_level))) {
-      warning("PSNU level ", psnu_level, " not included in model outputs.")
-    }
+  
+  if (is.null(psnu_level) || !psnu_level %in% naomi_output$meta_area$area_level) {
+    warning("PSNU level ", psnu_level, " not included in model outputs.")
```
[the remark this answers is not in the document]
Comment (jeffeaton) [reply R/pepfar-datapack.R:67]: Have reverted this back to just a normal R warning. Not worth the engineering to make it appear right now!

File: R/pepfar-datapack.R [hunk 3]
```diff
@@ -237,3 +264,28 @@ datapack_aggregate_1to9 <- function(indicators) {
   indicators[names(indicators1to9)] %>%
     dplyr::bind_rows(indicators1to9)
 }
+
+transform_dmppt2 <- function(x) {
+
+  if(any(names(x)[3:11] != c("area_id", "15-24", "25-34", "35-49", "50+",
+                             "15-24", "25-34", "35-49", "50+"))) {
+    stop("DMMPT2 output file does not have expected column names.")
+  }
```
[the remark this answers is not in the document]
Comment (r-ash) [reply R/pepfar-datapack.R:271]: Added that check, 

https://github.com/mrc-ide/hintr/pull/489/commits/bb86cd4064f3e0c1468c7605b129bc25f61a5c51

**Review by r-ash — APPROVED** [review]
This looks great

**Pull request merged**